data-model/src/main/scala/za/co/absa/enceladus/model/dataFrameFilter/package.scala

```diff
@@ -0,0 +1,98 @@
+  @JsonSubTypes(Array(
+    new Type(value = classOf[OrJoinedFilters], name = "OrJoinedFilters"),
+    new Type(value = classOf[AndJoinedFilters], name = "AndJoinedFilters"),
+    new Type(value = classOf[EqualFilter], name = "EqualFilter"),
+    new Type(value = classOf[DifferFilter], name = "DifferFilter")
+  ))
+  sealed trait DataFrameFilter {
+    @JsonIgnore def filter: Column
+    @JsonIgnore def or(filter: DataFrameFilter): DataFrameFilter = {
+      (this, filter) match {
+        case (a: OrJoinedFilters, b: OrJoinedFilters) => OrJoinedFilters(a.filterItems & b.filterItems)
+        case (a: OrJoinedFilters, b) => a.copy(filterItems = a.filterItems + b)
+        case (a, b: OrJoinedFilters) => b.copy(filterItems = b.filterItems + a)
+        case (a, b) => OrJoinedFilters(Set(a, b))
+      }
+    }
+
+    @JsonIgnore def and(filter: DataFrameFilter): DataFrameFilter = {
+      (this, filter) match {
+        case (a: AndJoinedFilters, b: AndJoinedFilters) => AndJoinedFilters(a.filterItems & b.filterItems)
+        case (a: AndJoinedFilters, b) => a.copy(filterItems = a.filterItems + b)
+        case (a, b: AndJoinedFilters) => b.copy(filterItems = b.filterItems + a)
+        case (a, b) => AndJoinedFilters(Set(a, b))
+      }
+    }
+
+    def + (filter: DataFrameFilter): DataFrameFilter = or(filter) //scalastyle:ignore class.name function used as operator
+    def *(filter: DataFrameFilter): DataFrameFilter = and(filter) //scalastyle:ignore class.name function used as operator
+
+  }
+
+  sealed trait JoinFilters extends DataFrameFilter {
+    @JsonIgnore protected def operator: (Column, Column) => Column
+
+    def filterItems: Set[DataFrameFilter]
+    def filter: Column = {
+      if (filterItems.isEmpty) {
+        lit(true)
+      } else {
+        filterItems.tail.foldLeft(filterItems.head.filter) { case(col, filterDef) =>
+          operator(col, filterDef.filter)
+        }
+      }
+    }
+  }
+
+  sealed trait SingleColumnAndValueFilter extends DataFrameFilter {
+    def columnName: String
+    def value: Any
+
+    @JsonIgnore protected def operator: (Column, Any) => Column
+    override def filter: Column = {
+      operator(col(columnName), value)
+    }
+  }
+
+  case class OrJoinedFilters(filterItems: Set[DataFrameFilter]) extends JoinFilters {
+    protected val operator: (Column, Column) => Column = (a: Column, b: Column) => { a or b }
+  }
+
+  case class AndJoinedFilters(filterItems: Set[DataFrameFilter]) extends JoinFilters {
+    protected val operator: (Column, Column) => Column = (a: Column, b: Column) => { a and b }
+  }
+
+  case class EqualFilter(columnName: String, value: Any) extends SingleColumnAndValueFilter {
+    protected val operator: (Column, Any) => Column = (column: Column, value: Any) => { column === value }
+  }
+
+  case class DifferFilter(columnName: String, value: Any) extends SingleColumnAndValueFilter {
```

Comment:
You might call me pedantic for this comment, but I'll take the risk :)
I would rename `EqualFilter` to `EqualsFilter` and `DifferFilter` to `NotEqualsFilter`. I did a quick google search, `EqualsFilter` yields much more search results than `EqualFilter` (8M vs 0.14M) from which I conclude that most people will be more used to `EqualsFilter`. Same for `NotEqualsFilter` vs. `DifferFilter` (2M vs 0.058M)

Reply:
I really like the `EqualsFilter` name.
I am not fond of the `NotEqualsFilter` though. `DiffersFilter` (with added _s_) feel more apt here. Others please, what's your opinion?

Reply:
I would also tend towards `EqualsFilter` and `DiffersFilter`, mostly because it's less verbose

Reply:
Fan of `NotEqualsFilter`, here.

WWJD? ⛪ 😆